File: version4j/src/main/java/com/only4/algorithm/version4j/leetcode/MoveZeroes.java
package com.only4.algorithm.version4j.leetcode;

/**
 * @author zhenyu.jiang
 */
public class MoveZeroes {
    public void moveZeroes(int[] nums) {
        int slow = 0; // 指向下一个非零元素应该放置的位置

        for (int fast = 0; fast < nums.length; fast++) {
            // 如果当前元素不为0，将其移动到slow位置
            if (nums[fast] != 0) {
                if (fast != slow) {
                    // 交换元素
                    int temp = nums[fast];
                    nums[fast] = nums[slow];
                    nums[slow] = temp;
                }
                slow++; // 更新下一个非零元素的位置
            }
        }
    }
}

File: version4k/src/main/kotlin/com/only4/algorithm/version4k/leetcode/moveZeroes.kt
package com.only4.algorithm.version4k.leetcode

fun moveZeroes(nums: IntArray) {
    var slow = 0 // 指向下一个非零元素应该放置的位置

    for (fast in nums.indices) {
        // 如果当前元素不为0，将其移动到slow位置
        if (nums[fast] != 0) {
            if (fast != slow) {
                // Kotlin的also语法实现交换
                nums[slow] = nums[fast].also { nums[fast] = nums[slow] }
            }
            slow++ // 更新下一个非零元素的位置
        }
    }
}
